libraries/ESP8266WiFi/src/BearSSLHelpers.cpp

```diff
@@ -0,0 +1,830 @@
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+typedef struct {
+  char *name;
+  unsigned char *data;
+  size_t data_len;
+} pem_object;
+
+// Forward definitions
+void free_ta_contents(br_x509_trust_anchor *ta);
+void free_public_key(public_key *pk);
+void free_private_key(private_key *sk);
+bool looks_like_DER(const unsigned char *buf, size_t len);
+pem_object *decode_pem(const void *src, size_t len, size_t *num);
+void free_pem_object_contents(pem_object *po);
+
+// Used as callback multiple places to append a string to a vector
+static void byte_vector_append(void *ctx, const void *buff, size_t len)
+{
+  std::vector<uint8_t> *vec = static_cast<std::vector<uint8_t>*>(ctx);
+  vec->reserve(vec->size() + len); // Allocate extra space all at once
+  for (size_t i=0; i < len; i++) {
+    vec->push_back(((uint8_t*)buff)[i]);
+  }
+}
+
+static bool certificate_to_trust_anchor_inner(br_x509_trust_anchor *ta, const br_x509_certificate *xc)
+{
+  std::unique_ptr<br_x509_decoder_context> dc(new br_x509_decoder_context); // auto-delete on exit
+  std::vector<uint8_t> vdn;
+  br_x509_pkey *pk;
+
+  // Clear everything
+  memset(ta, 0, sizeof(*ta));
+
+  br_x509_decoder_init(dc.get(), byte_vector_append, (void*)&vdn, 0, 0);
+  br_x509_decoder_push(dc.get(), xc->data, xc->data_len);
+  pk = br_x509_decoder_get_pkey(dc.get());
+  if (pk == nullptr) {
+    goto cleanup;
+  }
+
+  // Copy the raw certificate data
+  ta->dn.data = (uint8_t*)malloc(vdn.size());
+  if (!ta->dn.data) {
+    goto cleanup;
+  }
+  memcpy(ta->dn.data, &vdn[0], vdn.size());
+  ta->dn.len = vdn.size();
+  ta->flags = 0;
+  if (br_x509_decoder_isCA(dc.get())) {
+    ta->flags |= BR_X509_TA_CA;
+  }
+
+  // Extract the public key
+  switch (pk->key_type) {
+    case BR_KEYTYPE_RSA:
+      ta->pkey.key_type = BR_KEYTYPE_RSA;
+      ta->pkey.key.rsa.n = (uint8_t*)malloc(pk->key.rsa.nlen);
+      ta->pkey.key.rsa.e = (uint8_t*)malloc(pk->key.rsa.elen);
+      if ((ta->pkey.key.rsa.n == nullptr) || (ta->pkey.key.rsa.e == nullptr)) {
+        goto cleanup;
+      }
+      memcpy(ta->pkey.key.rsa.n, pk->key.rsa.n, pk->key.rsa.nlen);
+      ta->pkey.key.rsa.nlen = pk->key.rsa.nlen;
+      memcpy(ta->pkey.key.rsa.e, pk->key.rsa.e, pk->key.rsa.elen);
+      ta->pkey.key.rsa.elen = pk->key.rsa.elen;
+      return true;
+    case BR_KEYTYPE_EC:
+      ta->pkey.key_type = BR_KEYTYPE_EC;
+      ta->pkey.key.ec.curve = pk->key.ec.curve;
+      ta->pkey.key.ec.q = (uint8_t*)malloc(pk->key.ec.qlen);
+      if (ta->pkey.key.ec.q == nullptr) {
+        goto cleanup;
+      }
+      memcpy(ta->pkey.key.ec.q, pk->key.ec.q, pk->key.ec.qlen);
+      ta->pkey.key.ec.qlen = pk->key.ec.qlen;
+      return true;
+    default:
+      goto cleanup;
+  }
+  return true;
+
+  // In case of failure, free any allocations and fail out
+cleanup:
```

Comment:
I really dislike gotos in code. How about some scope guard that calls free_ta_contents(ta) when destroyed? I was looking into a std::unique_ptr with a custom deleter, maybe something like [this](https://codereview.stackexchange.com/questions/134234/on-the-fly-destructors)?